### ISSUE TITLE
20221130 normalizing report presence

### ISSUE THREAD

**Issue by robnagler** [radiasoft/sirepo#4818]
@garsuga sez:
> I was thinking about the best approach to making the "Start simulation" workflow and I was wondering if it was required that the reports' panels remain hidden until that button is pressed. Leaving the panels present and having them hide the graphs/visuals until data is ready would leave a lot less moving parts and would mean less defining things in the schema. That way each report could explain their 'urgency' (whether they are automatic or started by a button somewhere) in their local schema and the individual graphs would be shown/hidden depending on data availability instead of needing to have the panels be aware of these reports' completion.
>
> Added bonus is that the user could still access the configuration and forms present in the panels that have reports without running a simulation first

I think we should normalize this behavior. Either all graphs are hidden when they don't have simulation data or they are shown with blank data. This shouldn't be a per app/ad hoc decision.

As we discussed, we need some way of indicating a report is invalidated by the simulation data changing. I think this again relates to decoupling of "start" and "data available" and clear decision making about when a report is invalid or not. The server can then invalidate reports as soon as the data are invalid and all data stays consistent on the server (deleting run directories at the same time as a compute job is invalidated by the data).

@moellep what are your thoughts?

**Comment by robnagler** [comment]
> [report] knows it has data and needs to get it from the server

I think we should speak in different terms: the view registers interest in the report. This is in sirepo-data.json already[1]. I think instead of "panelState.hidden" we should have state associated with each panel (report). However, it's not clear to me that the model for a report can't be used by two panels. If that's the case, then we need to add a new model for this, or avoid this situation.

[1] Not for all apps save panelState, which is part of what we are trying to fix with the ngui project.

**Comment by garsuga** [comment]
I believe we are all talking on slightly different issues here. My concern is not about whether to hide the graph, as I think it only makes sense to hide the graph when it does not have data. I am suggesting we choose to show the panel that contains the graph,and optionally replace the graph with a placeholder stating it is not loaded yet or loading. In a code sense this is much less complex than also hiding the panel.

> Many apps have different reports depending on the simulation settings and it isn't possible to know which ones until the simulation is running.

This is not really a concern in this instance. Reports in the react version right now wait for data and show if it is present. They do not need to understand where it comes from. I am only worried about hiding/showing things here and I think it is mostly up to UI preference. I see leaving the panels behind as an advantage for a few reasons.

- Users can modify settings on the graphs / form elements in these panels without running the simulation first
- Users are aware of what data is not available by looking at which panels still need data.
- The implementation is simpler because panels do not need to coordinate with report data

I can see a possible issue though, when panels would typically be hidden because of settings on the simulation then having these panels be shown might be confusing / cause clutter.

In either case I think this issue is just a question of UI preference.

**Comment by garsuga** [comment]
I will finish up a draft of some behavior I think will be fine and have you try it out.

**Comment by robnagler** [comment]
@garsuga I set a deadline for this to 20221130. We need clarity on the approach you want to try.

**Comment by garsuga** [comment]
I think our current implementation is ideal. We do not have any cases that I have found yet where it would be more convenient to have access to forms in report panels before they have data generated.

**Comment by moellep** [comment]
Yes, this is the current behavior of the angularjs app as well.

**Comment by robnagler** [comment]
The angularjs implementation is not what I would call ideal. I think we want more consistency between manual and automatic graphs. I also do not like that we are showing reports that are out of date. They should disappear as soon as data changes and/or as soon as the server says the data has changed.

**Comment by moellep** [comment]
I think other items are much higher priority such as login, nav bar, etc. It is a tricky problem because you need to know which model data invalidates the reports - for example, changing a report model or simulation name doesn't invalidate all the reports.

**Comment by garsuga** [comment]
Yes, I agree. The server should probably maintain all of this information. It should mark reports as invalid server-side so the client does not need to duplicate this information

**Comment by robnagler** [comment]
> I think other items are much higher priority such as login, nav bar, etc. 

We can change the priority. Doesn't change the relevance of the issue.

> It is a tricky problem because you need to know which model data invalidates the reports - for example, changing a report model or simulation name doesn't invalidate all the reports.

The problem has to be solved on the server or client. Right now the client is making ad hoc decisions which do not hold on refresh. It's better solved on the server. If a simulation name changes, it doesn't have to change the parameter hash. It does change the serial, but that doesn't have to invalidate a report.

For example, in angular changing the report name causes automatic graphs to be re-requested. That seems incorrect.

**Comment by moellep** [comment]
Yes, I think this priority should be bumped until the react app approaches the features of the angularjs app.

**Comment by robnagler** [comment]
Some of the features are going to run into this same issue. For example, we should fix discard changes to example on the server, not the client. There's a balance here between features and engineering. We need to put more engineering. The pressure to get finished with the current set of React apps will be reduced after this week.
